manager/container.go

```diff
@@ -555,6 +558,8 @@ func (cd *containerData) housekeepingTick(timer <-chan time.Time, longHousekeepi
 		klog.V(3).Infof("[%s] Housekeeping took %s", cd.info.Name, duration)
 	}
 	cd.notifyOnDemand()
+	cd.lock.Lock()
```

Comment:
I believe that call above this line is responsible for synchronization with `containerData.OnDemandHousekeeping()`. Lock that you are trying to introduce looks redundant to me. Are you able to provide a test case that help race to manifest?

Reply:
I don't have a test case, but we were using cadvisor and our race detector tripped up:

```
WARNING: DATA RACE
Read at 0x00c0001e7a50 by goroutine 30:
  github.com/google/cadvisor/manager.(*containerData).OnDemandHousekeeping()
      vendor/github.com/google/cadvisor/manager/container.go:141 +0x9b
  github.com/google/cadvisor/manager.(*manager).getRequestedContainers.func1()
      vendor/github.com/google/cadvisor/manager/manager.go:741 +0x64

Previous write at 0x00c0001e7a50 by goroutine 130:
  github.com/google/cadvisor/manager.(*containerData).housekeepingTick()
      vendor/github.com/google/cadvisor/manager/container.go:558 +0x34c
  github.com/google/cadvisor/manager.(*containerData).housekeeping()
      vendor/github.com/google/cadvisor/manager/container.go:494 +0x304
```

Reply:
Thank you 🙇. I think we should use single synchronisation mechanism in this case. I wonder if moving notifyOnDemand() call a bit lower and applying similar change to OnDemandHousekeeping() would not solve the issue.

Reply:
Thanks for looking. I'm not quite sure that I follow what you mean by single synchronization. `OnDemandHousekeeping` is an external API so anyone can call it, and it can potentially race with `housekeepingTick` when both try and access `statsLastUpdatedTime`. Access to this variable should be guarded by some mutual exclusion - right?

Reply:
@iwankgb Can you take a look at this again? Thanks!

Reply:
@pkagrawal I don't think I will be able to look at this issue again before coming weekend.